### ISSUE TITLE
AWS Media Package with CastLabs DRMtoday - Unsupported keySystem or supportedConfigurations

### ISSUE THREAD

**Issue by nick-myers-dt** [videojs/videojs-contrib-eme#128]
Hello,

We've ingested, converted and packaged over 500 videos with AWS MediaConverter, AWS MediaPackager, SPEKE and CastLabs DRMtoday.

This has worked really well, however we're bumping into a couple of issues which we are unable to pinpoint the cause of.

A small number of these videos present an error message: "Unsupported keySystem or supportedConfigurations." in Edge and Chrome.

When we inspect the EME events, we see this sort of message:

```
RequestMediaKeySystemAccessCall Promise Result 22:42:54:195
    status:  rejected
    args:  [com.widevine.alpha, [
                audioCapabilities:  [
                        contentType:  audio/mp4; codecs="undefined"]
                videoCapabilities:  [
                        contentType:  video/mp4; codecs="avc1.640028"]]]
    result:  DOMException{}
```

```
RequestMediaKeySystemAccessCall Promise Result 22:42:54:196
    status:  rejected
    args:  [com.apple.fps.1_0, [
                audioCapabilities:  [
                        contentType:  audio/mp4; codecs="undefined"]
                videoCapabilities:  [
                        contentType:  video/mp4; codecs="avc1.640028"]]]
    result:  DOMException{}

```

We can also create this error by opening many tabs (each with a different video on them) in quick succession.  A reload of the page/player solves this.

Firefox states: "Key system is unsupported" with the full console error being:

`VIDEOJS: ERROR: (CODE:5 MEDIA_ERR_ENCRYPTED) Key system is unsupported `

Safari behaves as if it's playing, however no sound or video is present.

And we instantiate the player as follows:

```tsx
function getDashLicenseFor(assetId, authToken) {
  return (emeOptions, keyMessage, callback): void => {
    getWidevineDrmLicense({
      assetId,
      authToken,
      body: keyMessage,
      headers: emeOptions.emeHeaders,
    })
      .then((licence) => {
        callback(null, licence);
      })
      .catch((err) => {
        callback(err);
      });
  };
}

function getHlsLicenceFor(assetId, authToken) {
  return (emeOptions, contentId, keyMessage, callback): void => {
    getFairplayDrmLicense({
      assetId,
      authToken,
      body: keyMessage,
      headers: emeOptions.emeHeaders,
    })
      .then((licence) => {
        callback(null, licence);
      })
      .catch((err) => {
        callback(err);
      });
  };
}

const isHls = (videoUrl: string): boolean => {
  return videoUrl.endsWith('.m3u8');
};

const HLS_CONTENT_TYPE = 'application/x-mpegURL';
const DASH_CONTENT_TYPE = 'application/dash+xml';

const VideoPlayer: React.FunctionComponent<Props> = ({
  assetId,
  authToken,
  videoUrl,
  subtitles,
  poster,
  playCallback,
  waitCallback,
  controlBarChangeCallback,
  startTimestampInSeconds,
  endTimestampInSeconds,
}: Props) => {
  const [playerVisible, setPlayerVisible] = useState(false);
  const [isPlaying, setIsPlaying] = useState(false);
  const [isWaiting, setIsWaiting] = useState(true);
  const [controlBar, setControlBar] = useState<ControlBarEvent>({ controlBar: null });
  const [videoPlayer, setVideoPlayer] = useState<VideoJsPlayer | null>(null);
  const videoTag = useRef(null);
  const videoJsOptions: videojs.PlayerOptions = {
    controls: true,
    sources: [({
      src: videoUrl,
      type: isHls(videoUrl) ? HLS_CONTENT_TYPE : DASH_CONTENT_TYPE,
      poster,
      preload: 'none',
      keySystems: {
        'com.widevine.alpha': {
          getLicense: getDashLicenseFor(assetId, authToken),
        },
        'com.apple.fps.1_0': {
          getLicense: getHlsLicenceFor(assetId, authToken),
          certificateUri: DRM_TODAY_FAIRPLAY_CERTIFICATE_URL,
          getContentId: (): string => {
            return assetId;
          },
        },
      },
    } as SourceObject)],
    controlBar: {
      children: [
        'playToggle',
        'volumePanel',
        'currentTimeDisplay',
        'timeDivider',
        'durationDisplay',
        'progressControl',
        'descriptionsButton',
        'subtitlesButton',
        'captionsButton',
        'subsCapsButton',
        'audioTrackButton',
        'pictureInPictureToggle',
        'fullscreenToggle',
      ],
    },
  };
```

Any help or guidance would be most gratefully received.

**Comment by nick-myers-dt** [comment]
Replying to my own issue.  We're currently investigating whether the `contentType:  audio/mp4; codecs="undefined"]` is causing the issue, however we're unsure what we can do to rectify this particular failure.

**Comment by nick-myers-dt** [comment]
Closing this issue.  We determined that the issue was due to our configuration of AWS Media Converter which produced the last segment without any audio tracks, even though it contained some video, the segment was very short, 67ms. 

Media Package (at the time of writing) looks at last .ts segment produced by MediaConvert, if last segment contains no audio, MediaPackage will assume there is no valid audio tracks and will produce MPEG DASH playlist without it, which is what happened here.

MediaConvert offers an option to prevent this, the settings is called "Minimum final segment length"

When you specify a value for this setting, the encoder will combine any final segment that is shorter than the length that you specify with the previous segment. For example, if your segment length is 6 seconds and your final segment is 0.067 seconds without a minimum final segment length; when you set the minimum final segment length to 1, your final segment is 6.067 seconds

To mitigate this issue we re-transcoded non-playable assets with MediaConvert and re-ingested with MediaPackage and the problem was solved.
